### DISCUSSION TITLE
Generate l1_batch root_hash issues

### DISCUSSION THREAD

**Discussion by UrsaMajor-t** [zkSync-Community-Hub/zksync-developers#11] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

The root_hash of the known l1_batch is to record the state difference generated by the txs execution of the current batch, that is, one or more storage_log are generated after each tx execution, and all storage_logs under the current batch generate a root_hash.
That is to say, through storage_log, it can be proved that a certain tx exists in a certain batch.

### What is the specific issue or error you're encountering?

The question is, how do I get the storage_log. 
Ask from another angle:
Is there any connection between the storage_log generated by the transaction after zk_vm and the log generated by the transaction?
1. If there is a connection, how can logs be converted into storage_log?

2. If there is no connection, it is known that all storage_log are saved in the database, so I would like to ask if a storage_log with read-only permissions can be given to calculate root_hash.

### Can you share the error messages or logs you're receiving, if any?

none

### Have you made any recent changes to the contract before encountering this issue?

none

### Are there any external libraries or contracts that your contract interacts with?

none

### Can you provide the relevant portions of your contract code where the issue is occurring?

none

### Have you tried to isolate the problem, and if so, what were the results?

none

### What steps have you already taken to try to resolve the issue?

none

**Comment by bxpana** [comment, marked as answer]
Storage logs and txn logs are totally separate. Txn logs catch events for txns and Storage logs are for when the VM reads some slots of storage in a contract.

You can use the RPC method `getL1BatchDetails` (https://era.zksync.io/docs/api/js/providers.html#getl1batchdetails)

Right now you can't request storage logs via the API

**Reply by UrsaMajor-t** [reply]
in addition to api, are there other ways to obtain storage_log? in other words, can a storage_log database with read-only permission be opened, because this is very meaningful for transaction proof

**Reply by cc40513** [reply]
good

**Comment by mattbrathwaite** [comment]
To obtain the storage logs generated by transactions executed in zkSync, you can query the zkSync database where these logs are stored. However, direct access to storage logs may not be provided due to security and privacy concerns. Instead, consider leveraging zkSync's APIs or tools provided by zkSync developers to access relevant information about transactions and batches.

**Comment by propmacboorsp** [comment]
To obtain storage logs in the context of zkVM transactions, you can explore the event emission mechanism. Logs generated during zkVM transactions can be converted into storage logs through event emissions and event handling mechanisms. Ensure that your zkVM contracts emit events for relevant state changes, and then handle these events to construct storage logs. Alternatively, consider integrating logging mechanisms within your zkVM contracts to capture state changes directly.
